File: apps/nvim/lua/seismic/project_sync.lua
local M = {}

local config = require("seismic.config")
local detector = require("seismic.detector")

local last_sync = 0
local MIN_SYNC_INTERVAL_SECONDS = 15

local function send(payload)
	local api_key = config.get_api_key()
	local api_url = config.get_api_url()
	local body = vim.json.encode(payload)

	vim.system({
		"curl", "-s", "-o", "/dev/null",
		"-X", "POST",
		"-H", "Content-Type: application/json",
		"-H", "Authorization: Bearer " .. api_key,
		"-d", body,
		api_url .. "/api/projects/sync",
	}, { text = true })
end

function M.sync(force)
	if not config.is_enabled() or not config.has_api_key() then
		return
	end

	local now = os.time()
	if not force and now - last_sync < MIN_SYNC_INTERVAL_SECONDS then
		return
	end

	config.refresh_editor_settings()
	local metadata = detector.detect_project_metadata(config.use_git_root_project_name())
	if not metadata.repoUrl and #metadata.commits == 0 and not metadata.websiteUrl then
		return
	end

	last_sync = now
	send(metadata)
end

return M

File: apps/vscode/src/projectSync.ts
import * as vscode from 'vscode';
import * as config from './config';
import * as detector from './detector';

interface ProjectSyncPayload {
  project: string;
  repoUrl?: string;
  websiteUrl?: string;
  lastCommitAt?: number;
  commits: detector.GitCommit[];
}

const MIN_SYNC_INTERVAL_MS = 15 * 1000;

export class ProjectSyncService {
  private lastSyncByFolder = new Map<string, number>();
  private hasShownInvalidKeyWarning = false;

  async syncWorkspaceFolders(force = false): Promise<void> {
    const folders = vscode.workspace.workspaceFolders ?? [];
    await Promise.all(folders.map((folder) => this.syncWorkspaceFolder(folder, force)));
  }

  async syncDocument(document: vscode.TextDocument, force = false): Promise<void> {
    if (!config.isEnabled()) return;
    if (!detector.shouldTrack(document)) return;

    const folder = vscode.workspace.getWorkspaceFolder(document.uri);
    if (!folder) return;

    await this.syncWorkspaceFolder(folder, force);
  }

  async syncWorkspaceFolder(folder: vscode.WorkspaceFolder, force = false): Promise<void> {
    if (!config.isEnabled()) return;
    if (!config.hasApiKey()) return;

    const folderPath = folder.uri.fsPath;
    const lastSync = this.lastSyncByFolder.get(folderPath) ?? 0;
    if (!force && Date.now() - lastSync < MIN_SYNC_INTERVAL_MS) return;

    const metadata = await detector.detectWorkspaceProjectMetadata(folder);
    const payload: ProjectSyncPayload = {
      project: metadata.project ?? folder.name,
      repoUrl: metadata.repoUrl,
      websiteUrl: metadata.websiteUrl,
      lastCommitAt: metadata.lastCommitAt,
      commits: metadata.commits ?? [],
    };

    if (!payload.repoUrl && payload.commits.length === 0 && !payload.websiteUrl) return;

    const ok = await this.send(payload);
    if (ok) this.lastSyncByFolder.set(folderPath, Date.now());
  }

  private async send(payload: ProjectSyncPayload): Promise<boolean> {
    const apiKey = config.getApiKey();
    const apiUrl = config.getApiUrl();

    try {
      const res = await fetch(`${apiUrl}/api/projects/sync`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          Authorization: `Bearer ${apiKey}`,
        },
        body: JSON.stringify(payload),
      });

      if (res.status === 401) {
        this.notifyInvalidKey();
        return false;
      }

      return res.ok;
    } catch {
      return false;
    }
  }

  private notifyInvalidKey(): void {
    if (this.hasShownInvalidKeyWarning) return;
    this.hasShownInvalidKeyWarning = true;
    vscode.window.showWarningMessage(
      'Seismic: Invalid API key. Run "Seismic: Set API Key" to update it.',
    );
  }
}
